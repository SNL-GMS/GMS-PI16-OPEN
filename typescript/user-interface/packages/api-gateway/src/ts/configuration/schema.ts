import { readFileSync } from 'fs';
// eslint-disable-next-line import/no-extraneous-dependencies
import { resolve } from 'path';

/**
 * GraphQL schema definition for the configuration API gateway
 */

// GraphQL schema definitions
export const schema = readFileSync(
  resolve(process.cwd(), 'resources/graphql/configuration/schema.graphql')
).toString();
