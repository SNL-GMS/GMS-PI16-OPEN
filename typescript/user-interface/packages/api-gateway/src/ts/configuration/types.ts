import { ConfigurationTypes } from '@gms/common-model';
import Immutable from 'immutable';

/**
 * Overall configuration
 */
export interface Configuration {
  /**
   * AnalystConfiguration keyed on user role
   */
  analystConfiguration: Immutable.Map<string, ConfigurationTypes.ProcessingAnalystConfiguration>;

  /**
   * SohConfigurationOld
   */
  sohConfiguration: ConfigurationTypes.SohConfiguration;

  /**
   * CommonConfiguration keyed on user role
   */
  commonConfiguration: Immutable.Map<string, ConfigurationTypes.ProcessingCommonConfiguration>;
}

/**
 * Default user role
 */
export const defaultUserRole = 'DEFAULT';
