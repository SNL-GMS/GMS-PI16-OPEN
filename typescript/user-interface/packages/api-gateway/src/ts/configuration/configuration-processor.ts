import { ConfigurationTypes, WaveformTypes } from '@gms/common-model';
import { convertDurationToMilliseconds, sleep } from '@gms/common-util';
import { ApolloError } from 'apollo-server';
import config from 'config';
import Immutable from 'immutable';

import { gatewayLogger as logger } from '../log/gateway-logger';
import { HttpClientWrapper, HttpResponse } from '../util/http-wrapper';
import * as configurationMockBackend from './configuration-mock-backend';
import { Configuration, defaultUserRole } from './types';

/**
 * Singleton class used to retrieve Configuration attributes such as network and waveform filters
 * At start up a default network is fetched from the configuration service, if a user who logs in isn't the default
 * that users configuration will be fetched, at start up the station processors need a network and use a the default
 * network therefore the async method getDefaultNetwork is used, other wise after log in, can guarantee that
 * a users configuration will be loaded and the other get methods don't need to be awaited on.
 *
 */
export class ConfigurationProcessor {
  /**
   * Returns the singleton instance of the cache processor.
   *
   * @returns the instance of the cache processor
   */
  public static Instance(): ConfigurationProcessor {
    if (!ConfigurationProcessor.isInitialized) {
      ConfigurationProcessor.instance = new ConfigurationProcessor();
      ConfigurationProcessor.instance.initialize();
      ConfigurationProcessor.isInitialized = true;
    }
    return ConfigurationProcessor.instance;
  }

  /** The singleton instance */
  private static instance: ConfigurationProcessor;

  private static isInitialized = false;

  /**
   * The configuration
   */
  private readonly configuration: Configuration;

  /** Local configuration settings for reference stations */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private readonly settings: any;

  /** HTTP client wrapper for communicating with backend services */
  private readonly httpWrapper: HttpClientWrapper;

  /**
   * Constructor - initialize the processor, loading settings and initializing the HTTP client wrapper.
   */
  private constructor() {
    // Load configuration settings
    this.settings = config.get('configuration');

    // Initialize an http client
    this.httpWrapper = new HttpClientWrapper();

    // Initialize the configuration
    this.configuration = {
      analystConfiguration: Immutable.Map(),
      sohConfiguration: undefined,
      commonConfiguration: Immutable.Map()
    };
  }

  /**
   * Gets the analyst configuration based on the role
   *
   * @param role of the user from the user context
   * @returns a analyst configuration
   */
  public getAnalystConfiguration(
    role: string = defaultUserRole
  ): ConfigurationTypes.ProcessingAnalystConfiguration {
    if (this.configuration.analystConfiguration.has(role)) {
      return this.configuration.analystConfiguration.get(role);
    }
    throw new ApolloError(
      `Configuration error, could not find analyst configuration for role: ${role}`
    );
  }

  /**
   * Gets the common configuration based on the role
   *
   * @param role of the user from the user context
   * @returns a common configuration
   */
  public getCommonConfiguration(
    role: string = defaultUserRole
  ): ConfigurationTypes.ProcessingCommonConfiguration {
    if (this.configuration.commonConfiguration.has(role)) {
      return this.configuration.commonConfiguration.get(role);
    }
    throw new ApolloError(
      `Configuration error, could not find common configuration for role: ${role}`
    );
  }

  /**
   * Gets the network based on the role
   *
   * @param role of the user from the user context
   * @returns a network as a Promise<string>
   */
  public getNetworkForUser(role: string = defaultUserRole): string {
    if (this.configuration.analystConfiguration.has(role)) {
      return this.configuration.analystConfiguration.get(role).defaultNetwork;
    }
    throw new ApolloError(
      `Configuration error, could not find configuration default network for role: ${role}`
    );
  }

  /**
   * Gets the waveform filters based on the role
   *
   * @param role of the user from the user context
   * @returns a waveform filters as Promise<WaveformFilter[]>
   */
  public getWaveformFiltersForUser(role: string = defaultUserRole): WaveformTypes.WaveformFilter[] {
    if (this.configuration.analystConfiguration.has(role)) {
      return this.configuration.analystConfiguration.get(role).defaultFilters;
    }
    throw new ApolloError(
      `Configuration error, could not find configuration default filters for role: ${role}`
    );
  }

  /**
   * Gets the soh configuration based on the role
   *
   * @returns a soh configuration
   */
  public getSohConfiguration(): ConfigurationTypes.SohConfiguration {
    if (this.configuration.sohConfiguration) {
      return this.configuration.sohConfiguration;
    }
    throw new ApolloError(`Configuration error, could not find soh configuration`);
  }

  /**
   * Gets the soh configuration based on the role
   *
   * @returns a soh configuration
   */
  public getStationGroupNamesWithPriorities(): ConfigurationTypes.SOHStationGroupNameWithPriority[] {
    if (!this.configuration.sohConfiguration) {
      return [];
    }
    return this.configuration.sohConfiguration.stationSohControlConfiguration.displayedStationGroups.map(
      (sg, count) => ({ name: sg, priority: count })
    );
  }

  /**
   * Gets the SOH Update Interval based on the role
   *
   * @param role of the user from the user context
   * @returns the interval in seconds
   */
  public getSohUpdateIntervalForUser(role: string = defaultUserRole): number {
    if (this.configuration.sohConfiguration) {
      const sohParams = this.configuration.sohConfiguration.stationSohMonitoringDisplayParameters;
      return convertDurationToMilliseconds(sohParams.redisplayPeriod);
    }
    throw new ApolloError(
      `Configuration error, could not find configuration redisplayPeriod for role: ${role}`
    );
  }

  /**
   * Gets the SOH quiet interval based on the role
   *
   * @param role of the user from the user context
   * @returns the interval in seconds
   */
  public getSohDefaultQuietInterval(role: string = defaultUserRole): number {
    if (this.configuration.sohConfiguration) {
      const sohParams = this.configuration.sohConfiguration.stationSohMonitoringDisplayParameters;
      return convertDurationToMilliseconds(sohParams.acknowledgementQuietDuration);
    }
    throw new ApolloError(
      `Configuration error, could not find configuration acknowledgementQuietDuration for role: ${role}`
    );
  }

  /**
   * Load and fetch the initial configurations for the given user role
   *
   * @param role of the user from the user context
   */
  public async fetchConfiguration(): Promise<Configuration> {
    // if needed, fetch configurations
    await this.fetchAnalystConfiguration();
    await this.fetchSohConfiguration();
    await this.fetchCommonConfiguration();
    return this.configuration;
  }

  /**
   * Fetches AnalystConfiguration from the configuration service for the given unique configuration id
   *
   * @param configuration the configuration id to retrieve
   */
  private async fetchAnalystConfiguration(
    role: string = defaultUserRole
  ): Promise<ConfigurationTypes.ProcessingAnalystConfiguration> {
    if (this.configuration.analystConfiguration.has(role)) {
      return this.configuration.analystConfiguration.get(role);
    }

    // get the configuration id based on the user role
    const configuration: string = ConfigurationTypes.AnalystConfigs[role];

    logger.info(
      `Fetching analyst configuration data: role: ${role} configuration: ${configuration}`
    );
    const analystConfiguration = await this.internalFetchSubConfiguration<
      ConfigurationTypes.ProcessingAnalystConfiguration
    >(configuration);
    this.configuration.analystConfiguration = this.configuration.analystConfiguration.set(
      role,
      analystConfiguration
    );
    return this.configuration.analystConfiguration.get(role);
  }

  /**
   * Fetches CommonConfiguration from the configuration service for the given unique configuration id
   *
   * @param configuration the configuration id to retrieve
   */
  private async fetchCommonConfiguration(
    role: string = defaultUserRole
  ): Promise<ConfigurationTypes.ProcessingCommonConfiguration> {
    if (this.configuration.commonConfiguration.has(role)) {
      return this.configuration.commonConfiguration.get(role);
    }

    // get the configuration id based on the user role
    const configuration: string = ConfigurationTypes.CommonConfigs[role];

    logger.info(
      `Fetching common configuration data: role: ${role} configuration: ${configuration}`
    );
    const commonConfiguration = await this.internalFetchSubConfiguration<
      ConfigurationTypes.ProcessingCommonConfiguration
    >(configuration);
    this.configuration.commonConfiguration = this.configuration.commonConfiguration.set(
      role,
      commonConfiguration
    );
    return this.configuration.commonConfiguration.get(role);
  }

  /**
   * Fetches SohConfiguration from the configuration service for the given unique configuration id
   *
   * @param configuration the configuration to retrieve
   */
  private async fetchSohConfiguration(): Promise<ConfigurationTypes.SohConfiguration> {
    // add data to the post because our service needs something in the body
    const postBodyData = 'test';
    // Retrieve the request configuration for the service call
    const requestConfiguration = this.settings.backend.services.getSohConfiguration.requestConfig;
    // Call the service and process the response data and add configuration to map
    const sohConfiguration = await this.callFetchSohConfiguration(
      requestConfiguration,
      postBodyData
    );
    this.configuration.sohConfiguration = sohConfiguration;
    return this.configuration.sohConfiguration;
  }

  /**
   * calls the webservice to get the sohconfiguration
   *
   * @param requestConfiguration configuration given to the http request wrapper
   * @param postBodyData string to add to the boyd of the request,
   * the call needs a string in the body of the request
   */
  private readonly callFetchSohConfiguration = async (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    requestConfiguration: any,
    postBodyData: string
  ): Promise<ConfigurationTypes.SohConfiguration> =>
    this.httpWrapper
      .request<ConfigurationTypes.SohConfiguration>(requestConfiguration, postBodyData)
      .then((response: HttpResponse<ConfigurationTypes.SohConfiguration>) => response.data)
      .catch(async e => {
        logger.warn(`Error failed to connect to soh configuration service: ${e}`);
        const timeoutMS = 2000;
        return sleep<Promise<ConfigurationTypes.SohConfiguration>>(timeoutMS).then(async () => {
          logger.info(`Retrying to connect to configuration service`);
          logger.info(`retry config:: ${JSON.stringify(requestConfiguration)}`);
          return this.callFetchSohConfiguration(requestConfiguration, postBodyData);
        });
      });

  /**
   * Hits configuration service to get back an AnalystConfiguration based on role
   *
   * @param configuration configuration to retrieve
   */
  private async internalFetchSubConfiguration<T>(
    configuration: string,
    selectors: ConfigurationTypes.Selector[] = []
  ): Promise<T> {
    // Build the query to be encoded as query string parameters
    const query = {
      configName: configuration,
      selectors
    };

    // Retrieve the request configuration for the service call
    const { requestConfig } = this.settings.backend.services.getAnalystConfiguration;

    logger.debug(
      `Requesting Analyst configuration query ${JSON.stringify(query)} config ${JSON.stringify(
        requestConfig
      )}`
    );

    // Call the service and process the response data and add configuration to map
    return this.httpWrapper
      .request<T>(requestConfig, query)
      .then((response: HttpResponse<T>) => response.data)
      .catch(async e => {
        logger.warn(`Error failed to connect to configuration service: ${e}`);
        const timeoutMS = 2000;

        return sleep<Promise<ConfigurationTypes.ProcessingAnalystConfiguration>>(timeoutMS).then(
          async () => {
            logger.info(`Retrying to connect to configuration service`);
            return this.internalFetchSubConfiguration<T>(configuration, selectors);
          }
        );
      });
  }

  /**
   * Initialize the configuration processor, fetching configuration from the backend.
   * This function sets up a mock backend if configured to do so.
   */
  private initialize() {
    logger.info(
      'Initializing the configuration processor - Mock Enable: %s',
      this.settings.backend.mock.enable
    );

    // If service mocking is enabled, initialize the mock backend
    if (this.settings.backend.mock.enable) {
      configurationMockBackend.initialize(this.httpWrapper.createHttpMockWrapper());
    }
  }
}
