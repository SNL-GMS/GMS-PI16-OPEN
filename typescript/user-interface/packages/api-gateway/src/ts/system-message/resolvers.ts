// GraphQL Resolvers
import { SystemMessageTypes } from '@gms/common-model';
import { MILLISECONDS_IN_SECOND, toEpochSeconds } from '@gms/common-util';
import config from 'config';

import { UserContext } from '../cache/types';
import { gatewayLogger as logger } from '../log/gateway-logger';
import { SystemMessageProcessor } from './system-message-processor';

// Load configuration settings
const settings = config.get('systemMessage');

export const resolvers = {
  // Query Resolvers
  Query: {
    // Returns System Message Definitions
    systemMessageDefinitions: async (
      // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
      _,
      // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
      __,
      userContext: UserContext
    ): Promise<SystemMessageTypes.SystemMessageDefinition[]> => {
      logger.info(`Getting System Message Definitions. User: ${userContext.userName}`);
      return SystemMessageProcessor.Instance().getSystemMessageDefinitions();
    }
  },
  // Mutation Resolvers
  Mutation: {},
  // Subscription Resolvers
  Subscription: {
    // the system message subscription resolver
    systemMessages: {
      // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
      subscribe: () =>
        SystemMessageProcessor.Instance().pubsub.asyncIterator(
          settings.subscriptions.systemMessages
        )
    }
  },
  SystemMessage: {
    time: (systemMessage): number => toEpochSeconds(systemMessage.time) * MILLISECONDS_IN_SECOND
  }
};
