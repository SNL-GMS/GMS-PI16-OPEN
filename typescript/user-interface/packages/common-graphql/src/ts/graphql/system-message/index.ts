import * as SystemMessageGqls from './gqls';
import * as SystemMessageQueries from './queries';
import * as SystemMessageSubscriptions from './subscriptions';

export { SystemMessageGqls, SystemMessageQueries, SystemMessageSubscriptions };
