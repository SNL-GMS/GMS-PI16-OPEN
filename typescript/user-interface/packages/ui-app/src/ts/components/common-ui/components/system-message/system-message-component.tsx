/* eslint-disable react/prop-types */
import { useForceGlUpdateOnResizeAndShow } from '@gms/ui-util';
import * as React from 'react';

import { Queries } from '~components/client-interface';
import { commonClassNames } from '~components/common-ui/config/common-class-names';

import { BaseDisplay } from '../base-display';
import { SystemMessageAlertPlayer } from './audio';
import { AudibleNotificationContext } from './audio/audible-notification-context';
import {
  validateAvailableSounds,
  validateConfiguredAudibleNotifications
} from './sound-configuration/sound-configuration-util';
import { SystemMessageTable } from './system-message-table';
import { useSeverityFilters } from './toolbar/severity-filters';
import { SystemMessageToolbar } from './toolbar/system-message-toolbar';
import { SystemMessageProps } from './types';

/**
 * The system message component
 *
 * @param props the system message component props
 */
// eslint-disable-next-line react/display-name
export const SystemMessage: React.FunctionComponent<SystemMessageProps> = React.memo(props => {
  const [severityFilterMap, setSeverityFilterMap] = useSeverityFilters();

  /** state for enabling and disabling auto scrolling */
  const [isAutoScrollingEnabled, setIsAutoScrollingEnabled] = React.useState(true);

  /** state for enabling and disabling sounds */
  const [isSoundEnabled, setIsSoundEnabled] = React.useState(true);

  /** validate available sound files; toast error message for each sound file that missing */
  React.useEffect(validateAvailableSounds, []);

  const userProfileQuery = Queries.UserProfileQuery.useUserProfileQuery();
  const userProfile = userProfileQuery?.data ?? undefined;
  const systemMessageDefinitionQuery = Queries.SystemMessageDefinitionQuery.useSystemMessageDefinitionQuery();
  const systemMessageDefinitions = systemMessageDefinitionQuery?.data ?? undefined;

  /** validate the configured audible notifications; toast error message for each sound file that is missing */
  React.useEffect(() => validateConfiguredAudibleNotifications(userProfile?.audibleNotifications), [
    userProfile?.audibleNotifications
  ]);

  /** force update on golden layout resize and show -> ensures that the toolbar is properly sized */
  useForceGlUpdateOnResizeAndShow(props.glContainer);

  return (
    <AudibleNotificationContext.Provider
      value={{
        audibleNotifications: userProfile?.audibleNotifications,
        setAudibleNotifications: props.setAudibleNotifications
      }}
    >
      <BaseDisplay
        glContainer={props.glContainer}
        className={`system-message-display ${commonClassNames.sharedTableClasses}`}
      >
        <SystemMessageToolbar
          /* eslint-disable @typescript-eslint/unbound-method */
          addSystemMessages={props.addSystemMessages}
          clearAllSystemMessages={props.clearAllSystemMessages}
          clearExpiredSystemMessages={props.clearExpiredSystemMessages}
          clearSystemMessages={props.clearSystemMessages}
          /* eslint-enable @typescript-eslint/unbound-method */
          systemMessagesState={props.systemMessagesState}
          isAutoScrollingEnabled={isAutoScrollingEnabled}
          setIsAutoScrollingEnabled={setIsAutoScrollingEnabled}
          isSoundEnabled={isSoundEnabled}
          setIsSoundEnabled={setIsSoundEnabled}
          systemMessageDefinitions={systemMessageDefinitions}
          severityFilterMap={severityFilterMap}
          setSeverityFilterMap={m => {
            setSeverityFilterMap(m);
          }}
        />
        <SystemMessageTable
          /* eslint-disable @typescript-eslint/unbound-method */
          addSystemMessages={props.addSystemMessages}
          clearAllSystemMessages={props.clearAllSystemMessages}
          clearExpiredSystemMessages={props.clearExpiredSystemMessages}
          clearSystemMessages={props.clearSystemMessages}
          /* eslint-enable @typescript-eslint/unbound-method */
          systemMessages={props.systemMessagesState.systemMessages?.filter(
            msg => severityFilterMap?.get(msg.severity) ?? true
          )}
          isAutoScrollingEnabled={isAutoScrollingEnabled}
          setIsAutoScrollingEnabled={setIsAutoScrollingEnabled}
          severityFilterMap={severityFilterMap}
        />
        <SystemMessageAlertPlayer
          isSoundEnabled={isSoundEnabled}
          latestSystemMessages={props.systemMessagesState.latestSystemMessages}
        />
      </BaseDisplay>
      )
    </AudibleNotificationContext.Provider>
  );
});
