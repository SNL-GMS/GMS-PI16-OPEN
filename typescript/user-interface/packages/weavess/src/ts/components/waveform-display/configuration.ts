import { WeavessConstants, WeavessTypes } from '@gms/weavess-core';
import defaultsDeep from 'lodash/defaultsDeep';
import memoizeOne from 'memoize-one';

/** Defines the default configuration for Weavess */
export const defaultConfiguration: WeavessTypes.Configuration = {
  defaultChannelHeightPx: WeavessConstants.DEFAULT_CHANNEL_HEIGHT_PIXELS,

  labelWidthPx: WeavessConstants.DEFAULT_LABEL_WIDTH_PIXELS,

  xAxisLabel: undefined,

  shouldRenderWaveforms: true,

  shouldRenderSpectrograms: true,

  hotKeys: {
    drawMeasureWindow: 'Alt',
    amplitudeScale: 'KeyS',
    amplitudeScaleSingleReset: 'Alt+KeyS',
    amplitudeScaleReset: 'Alt+Shift+KeyS',
    maskCreate: 'KeyM'
  },

  defaultChannel: {
    disableMeasureWindow: false,
    disableSignalDetectionModification: false,
    disablePreditedPhaseModification: false,
    disableMaskModification: false
  },

  nonDefaultChannel: {
    disableMeasureWindow: false,
    disableSignalDetectionModification: false,
    disablePreditedPhaseModification: false,
    disableMaskModification: false
  },

  colorScale: undefined
};

/**
 * Returns the Weavess configuration based on the configuration
 * passed in by the user and the default configuration
 *
 * @param config
 * @param defaultConfig
 */
// eslint-disable-next-line
const getConfiguration = (
  config: Partial<WeavessTypes.Configuration> | undefined,
  defaultConfig: WeavessTypes.Configuration = defaultConfiguration
): WeavessTypes.Configuration => defaultsDeep(config, defaultConfig);

export const memoizedGetConfiguration = memoizeOne(getConfiguration);
