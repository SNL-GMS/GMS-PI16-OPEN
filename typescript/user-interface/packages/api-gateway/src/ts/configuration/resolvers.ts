import { ConfigurationTypes } from '@gms/common-model';
import { convertDurationToMilliseconds, convertDurationToSeconds } from '@gms/common-util';

import { getLogLevel } from '../log/gateway-logger';
import { ConfigurationProcessor } from './configuration-processor';

// GraphQL Resolvers
export const resolvers = {
  // Query resolvers
  Query: {
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    uiAnalystConfiguration: () => {
      const logLevel = getLogLevel();
      // disable because the linter will automatically reformat this line to be too long
      /* eslint-disable max-len */
      const analystConfiguration: ConfigurationTypes.ProcessingAnalystConfiguration = ConfigurationProcessor.Instance().getAnalystConfiguration();
      const sohConfiguration: ConfigurationTypes.SohConfiguration = ConfigurationProcessor.Instance().getSohConfiguration();
      const commonConfiguration: ConfigurationTypes.ProcessingCommonConfiguration = ConfigurationProcessor.Instance().getCommonConfiguration();
      const sohParams = sohConfiguration.stationSohMonitoringDisplayParameters;
      /* eslint-enable max-len */

      return {
        logLevel,
        defaultNetwork: analystConfiguration.defaultNetwork,
        defaultFilters: analystConfiguration.defaultFilters,
        sohStationGroupNames: ConfigurationProcessor.Instance().getStationGroupNamesWithPriorities(),
        redisplayPeriod: convertDurationToMilliseconds(sohParams.redisplayPeriod),
        reprocessingPeriod: convertDurationToSeconds(
          sohConfiguration.stationSohControlConfiguration.reprocessingPeriod
        ),
        acknowledgementQuietDuration: convertDurationToMilliseconds(
          sohParams.acknowledgementQuietDuration
        ),
        availableQuietDurations: sohParams.availableQuietDurations.map(
          convertDurationToMilliseconds
        ),
        sohStationStaleTimeMS: convertDurationToMilliseconds(sohParams.sohStationStaleDuration),
        sohHistoricalDurations: sohParams.sohHistoricalDurations.map(convertDurationToMilliseconds),
        systemMessageLimit: commonConfiguration.systemMessageLimit
      };
    }
  }
};
