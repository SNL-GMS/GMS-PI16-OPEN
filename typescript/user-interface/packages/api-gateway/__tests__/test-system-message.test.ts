import { SystemMessageTypes } from '@gms/common-model';
import * as Immutable from 'immutable';

import { ConfigurationProcessor } from '../src/ts/configuration/configuration-processor';
import { SystemMessageProcessor } from '../src/ts/system-message/system-message-processor';
import { systemMessages } from './__data__/system-message-data';

/**
 * Sets up test by loading Configuration
 */
async function setupTest() {
  // Simulates an initial login request to fetch configuration
  await ConfigurationProcessor.Instance().fetchConfiguration();
}

beforeAll(async () => setupTest());

const systemMessageProcessor = SystemMessageProcessor.Instance();

describe('System Message Tests', () => {
  it('System Message Processor is defined', () => expect(systemMessageProcessor).toBeDefined());
  it('System Messages loaded', () => {
    expect(systemMessages).toBeDefined();
    expect(systemMessages.length).toBeGreaterThan(0);
  });
  it('Retrieve systemMessageDefinitions', async () => {
    const systemMessageDefs = await systemMessageProcessor.getSystemMessageDefinitions();
    expect(systemMessageDefs).toMatchSnapshot();
  });

  it('System Message Consume', () => {
    const immutableSystemMessages: Immutable.List<SystemMessageTypes.SystemMessage> = Immutable.List(
      systemMessages
    );
    systemMessageProcessor.registerKafkaConsumerCallbacks();
    systemMessageProcessor.consumeSystemMessages('system.system-messages', immutableSystemMessages);
    expect(true).toBeTruthy();
  });

  it('System Message Consume with undefined topic', () => {
    const immutableSystemMessages: Immutable.List<SystemMessageTypes.SystemMessage> = Immutable.List(
      systemMessages
    );
    systemMessageProcessor.registerKafkaConsumerCallbacks();
    systemMessageProcessor.consumeSystemMessages(undefined, immutableSystemMessages);
    expect(true).toBeTruthy();
  });
});
