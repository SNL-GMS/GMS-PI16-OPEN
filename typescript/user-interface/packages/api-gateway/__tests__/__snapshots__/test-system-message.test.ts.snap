// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`System Message Tests Retrieve systemMessageDefinitions 1`] = `
Array [
  Object {
    "systemMessageCategory": "SOH",
    "systemMessageSeverity": "INFO",
    "systemMessageSubCategory": "USER",
    "systemMessageType": "CHANNEL_MONITOR_TYPE_QUIET_PERIOD_CANCELED",
    "template": "Station %s Channel %s %s quiet period canceled by user %s",
  },
  Object {
    "systemMessageCategory": "SOH",
    "systemMessageSeverity": "INFO",
    "systemMessageSubCategory": "USER",
    "systemMessageType": "CHANNEL_MONITOR_TYPE_STATUS_CHANGE_ACKNOWLEDGED",
    "template": "Station %s Channel %s %s status change acknowledged by user %s|Station %s Channel %s %s status change acknowledged by user %s with comment '%s'",
  },
  Object {
    "systemMessageCategory": "SOH",
    "systemMessageSeverity": "INFO",
    "systemMessageSubCategory": "STATION",
    "systemMessageType": "STATION_SOH_STATUS_CHANGED",
    "template": "Station %s SOH status changed from %s to %s",
  },
  Object {
    "systemMessageCategory": "SOH",
    "systemMessageSeverity": "INFO",
    "systemMessageSubCategory": "STATION",
    "systemMessageType": "CHANNEL_MONITOR_TYPE_QUIET_PERIOD_EXPIRED",
    "template": "Station %s Channel %s %s quiet period expired",
  },
  Object {
    "systemMessageCategory": "SOH",
    "systemMessageSeverity": "WARNING",
    "systemMessageSubCategory": "CAPABILITY",
    "systemMessageType": "STATION_GROUP_CAPABILITY_STATUS_CHANGED",
    "template": "Station Group %s capability status changed from %s to %s",
  },
  Object {
    "systemMessageCategory": "SOH",
    "systemMessageSeverity": "INFO",
    "systemMessageSubCategory": "STATION",
    "systemMessageType": "CHANNEL_MONITOR_TYPE_STATUS_CHANGED",
    "template": "Station %s Channel %s %s status changed from %s (%s) to %s (%s)",
  },
  Object {
    "systemMessageCategory": "SOH",
    "systemMessageSeverity": "WARNING",
    "systemMessageSubCategory": "USER",
    "systemMessageType": "CHANNEL_MONITOR_TYPE_QUIETED",
    "template": "Station %s Channel %s %s quieted for %s by user %s|Station %s Channel %s %s quieted for %s by user %s with comment '%s'",
  },
  Object {
    "systemMessageCategory": "SOH",
    "systemMessageSeverity": "WARNING",
    "systemMessageSubCategory": "CAPABILITY",
    "systemMessageType": "STATION_CAPABILITY_STATUS_CHANGED",
    "template": "Station %s capability status for Station Group %s changed from %s to %s",
  },
  Object {
    "systemMessageCategory": "SOH",
    "systemMessageSeverity": "CRITICAL",
    "systemMessageSubCategory": "STATION",
    "systemMessageType": "STATION_NEEDS_ATTENTION",
    "template": "Station %s needs attention",
  },
]
`;
